### ISSUE TITLE
Should Chapel support promoted on-clauses?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#18980]
While working on slides today, I found myself wondering whether Chapel should support
promoted on-clauses as follows:

```chapel
on Locales[0..<numLocales-2] do ...
```

which we might imagine to most literally be interpreted as:

```chapel
forall loc in Locales[0..<numLocales/2] do
  on loc do ...
```

but we'd probably prefer to define it as:

```chapel
coforall loc in Locales[0..<numLocales/2] do
  on loc do ...
```

To the best of my memory, this hasn't been kicked around before.  But is
there any reason not to consider it?  It makes explicit parallelism implicit,
but in that sense is not so different from other forms of promotion.

**Comment by gbtitus** [comment]
Could be a fair amount of fun defining the precise semantics. Is this specific to on-stmts only on the `Locales` array, or on-stmts on any array? In any case, I'd think we'd want only one coforall-task for each locale holding any element of the array. I'd be a bit worried about the surprise! effect of promoting it to a coforall-stmt with as many tasks as there were distinct blocks, in the case where the array was distributed cyclic or block-cyclic and the blocksize was much smaller than the array size divided by the number of locales, i.e., each locale held many blocks of the array and there was a coforall-task for each of those blocks. But in general I like the idea.

**Comment by bradcray** [comment]
> Is this specific to on-stmts only on the Locales array, or on-stmts on any array?

Oh, that's a really interesting question.  I was thinking quickly from the perspective of "on-clauses take locales, so rather than a single locale, what if we used multiple locales?"  But of course you're right that on-clauses can take any expression with storage (and maybe even some without?), and that if we did a coforall per array element (say), that would be killer.  So maybe rather than interpreting:

```chapel
on someCollectionOfValues ...
```

as

```chapel
coforall v in someCollectionOfValues do
  on v do ...
```

we interpret it more like:

```chapel
coforall loc in someCollectionOfValues.targetLocales() do
  on loc do ...
```

?

**Comment by gbtitus** [comment]
Combining responses to a couple of things here ...

> ... interpret it more like:
> ```chapel
> coforall loc in someCollectionOfValues.targetLocales() do
> ```

Yes, I think I like that better.

> So maybe that suggests that if we pursued this, we'd only "promote" in the case of a collection of locales argument (and if the user wanted the "go to all target locales of a distributed array" behavior, they could say `on myBlockArr.targetLocales()` to get that.

Or maybe distinguish between `on <anArray> ...` and `on <aSliceOfAnArray>`, with the former doing a serial on-stmt to where the array was declared and the latter doing a parallel coforall-on-stmt to the `targetLocales` of the slice. Or is that too subtle? (Or even nonsensical; I haven't thought about this very long.)

**Comment by bradcray** [comment]
> Or maybe distinguish between on <anArray> ... and on <aSliceOfAnArray>, with the former doing a serial on-stmt to where the array was declared and the latter doing a parallel coforall-on-stmt to the targetLocales of the slice. Or is that too subtle? 

I think "too subtle."  In most contexts, we want array slices to behave as similarly to "normal" arrays as possible.  E.g., it'd be a shame if:

```chapel
on A[D] ...
```

behaved differently from:

```chapel
var B = A[D];
on B ...
```

**Comment by e-kayrakli** [comment]
> Hmm, yeah, good point (though note that that would be identical to the .targetLocales() behavior, right? So maybe the real question is "For a Block-distributed array, do you want the on to go to the locale where it happened to be declared, or to go to a multiplicity of locales. Arguable tradeoffs there are that the former has never seemed very useful / a bit arbitrary / and sometimes surprising/wrong. But the new interpretation could be surprising from the "introduces parallelism" perspective.
> 
> So maybe that suggests that if we pursued this, we'd only "promote" in the case of a collection of locales argument (and if the user wanted the "go to all target locales of a distributed array" behavior, they could say on myBlockArr.targetLocales() to get that.

I think this direction would be more OK for me if `myBlockArr.locale` would give us its `targetLocales`. Which I don't think we should do, and in the same vein, I find where we are today to be more appealing than what this proposal suggests.

I feel that every symbol having a `.locale` is a good place to be. I understand that that definition gets a bit vague for distributed arrays. But that's probably only because we know about privatization. If we move to a world where we just use serialization+RVF for distributed arrays, then, distributed arrays will have a more concrete `.locale`.

Put it differently, how are we supposed to support this for arbitrary data collections? Do we ask them to define a targetLocales? And if not, we'll go to the home locale of the symbol?

We can special case Chapel arrays to support this but then promoted `on`s will be just a way to sugar

```chapel
coforall l in someLocaleIteratorOrArray do on l {
```

Which I find to be sweet enough and does not need more sugar, personally :)

**Comment by bradcray** [comment]
> I understand that that definition gets a bit vague for distributed arrays.

To be clear, my motivation for proposing this wasn't that I'm particularly worried about how distributed arrays behave today (though they are a bit odd), but rather because I found myself typing `on Locales[0..<numLocales by 2]` on a slide as a means of trying to kick off a bunch of tasks on a number of locales before catching myself and realizing it wasn't actually legal.  Then I asked myself "But why isn't it?"

> Put it differently, how are we supposed to support this for arbitrary data collections? Do we ask them to define a targetLocales?

In the current evolution of the proposal (which I think is identical to the OP), I don't think we're asking or needing anything from arbitrary data collections beyond what they want to provide.  Specifically, I view the current proposal as:

1) `on myLocaleExpr` : works like today
2) `on myCollectionOfLocalesThatPromotes`: works as proposed here (where "that promotes" means "would be legal to pass to an argument `l: locale`)
3) `on otherExpr`: works like today (i.e., changes into `on otherExpr.locale`

My use of `.targetLocales()` in [the example above](https://github.com/chapel-lang/chapel/issues/18980#issuecomment-1011639393) wasn't meant to be a requirement that anything support that routine, simply that it is a reasonable way to explicitly get an array of locales from an array if one wanted to do this that didn't rely on any special treatment of arrays in on-clauses.

So, playing devil's advocate, I could imagine counterarguments to the proposal above as being:

* `on myArrayOfLocales` is too confusing because how would I know whether it's case 2 vs. case 3, which is what would happen for `on myArrayOfNonLocales`
 
I think that this is a reasonable concern and reason to hesitate, but I also think that it's arguable that the signature for an on-clause is effectively `proc on(l: locale)` and that treating this like promotion and `on(myNonLocaleExpr)` as being a coercion (where the implicit `.locale` implements the coercion) would be a reasonable counterargument for going with the proposal—`on`s expect locales, so providing a multiplicity of locales is just giving it more than one.


* `on myArrayOfLocales` being too subtle given that it's not _so_ terrible to write it the way we would have to today (with an accompanying `coforall`)

I also think this is reasonable, but I'm not sure it's more subtle than any other promotion, and I'd argue that, given the conceptual signature of the on-clause from the previous bullet, promotion is a more natural / obvious interpretation than "Go to the locale on which this array of locales happened to be declared."  Put another way, if you taught users about `on Locales[3]` and then you showed them `on Locales[0..<numLocales by 2]` I think they'd guess the behavior to be something like this proposal rather than something like `on Locales[0..<numLocales by 2].locale` (which is effectively `on here` since the slice was created here?).

Are there other counterarguments I'm missing?

**Comment by mppf** [comment]
I think the reason I don't like the proposal is that in makes `on` behave in a fundamentally different manner depending on the type of the argument to it. But I don't think of it as just "too confusing".

Say I am looking at some generic code and it has `on arg` where `arg` was an argument passed to it that is totally generic. How do I know what this will do?

Now, arguably we already have this problem due to the way `on someLocale` or `on someClassInstance` behave differently from `on myInteger`, say. But, these differences are limited to how you figure out which locale the `on` statement is going to go to. It is not nearly so huge as turning the `on` statement into a loop.

While I can see, from a PL theory kind of view point, that `on` is a function, it is not expressed that way syntactically. I think of it as a *sequential* control flow construct. So it would annoy me if it had the ability to introduce parallelism. And we have other sequential control flow constructs that arguably take in a single "argument" and you could say the same thing about them.

As one particular example: I could argue that `if` actually is a function that accepts a `bool` argument. And therefore `if myBoolArray { body() }` should run `body()` in a promoted manner, in parallel, once for each array element. To me, this makes just as much sense as the `on` proposal in this issue.

And yes, I could argue against promotion at all, but for a promoted call like `f(MyArray)`, when I am looking at the call site, I already have the problem that I don't necessarily know what `f` does. I think that is a fundamentally different situation from something like `if`.

**Comment by gbtitus** [comment]
> I think of [an on-stmt] as a _sequential_ control flow construct. So it would annoy me if it had the ability to introduce parallelism.

This is a very strong point; I totally agree, and am a tad embarrassed it didn't occur to me!

**Comment by bradcray** [comment]
I'm not necessarily convinced by the counterarguments so far, but I'm also not particularly dying for this feature, so unless Lydia wants to turn her thumbs-up into an argument that convinces people here, I'll close this as something that we're not going to do now, but could reconsider in the future if/when someone wanted it.

To Michael's comment about `on` being a serial concept, I'm not sure I buy it in the same way that `sin()` is a sequential construct until I pass an array to it (and, sure, maybe I don't know what sin() does... except that I do well enough to understand what's going on).  

I also don't necessarily buy that the argument that since it's not literally a function we shouldn't permit it to be promoted when a multiplicity of locales are supplied to  it.  Using Michael's promoted conditional example, I've found myself frequently questioning why we don't permit conditionals to be promoted (several times during Advent of Code, in fact), and had been planning to open a separate issue about that.  Why wouldn't `var B = if (A > 0) then 1 else -1;` give me an array `B` whose size/shape was the same as A, and whose elements reflected whether A was positive or negative (if I know A isn't ever zero, say)?  What is gained by requiring me to rewrite this in a less succinct way?  I've forked that question off to a separate issue: https://github.com/chapel-lang/chapel/issues/19068

**Comment by mppf** [comment]
FWIW I think the situation with a conditional expression is very different from a conditional statement.

I.e. I would be OK with this being promoted
``` chapel
var B = if A > 0 then 1 else -1;
```

but not this
``` chapel
if A > 0 {
  writeln("1");
} else {
  writeln("0");
}
```

Since we don't have `on` expressions, I view the `on` statement case as more similar to the 2nd of these two.
